### PR TITLE
Metadata crud endpoints

### PR DESCRIPTION
**Pull request by solomon-negusse**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [X] Check the commit's or even all commits' message styles matches our requested structure.
- [X] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Metadata are not exposed in the API 
Issue Number: [GTC-1347](https://gfw.atlassian.net/browse/GTC-1347)

## What is the new behavior?

Adds metadata crud and endpoints

## Does this introduce a breaking change?

- [] Yes
- [X] No

### CODE REVIEW

**Review by gtempus — COMMENTED** [review]
I added a few small comments in here, @solomon-negusse. This PR isn't as big as it looks because of all the other merged commits from other efforts. 😅 

I see that the tests are failing, still. Have you been doing something to get them to pass in your environment?
The error is : `ModuleNotFoundError: No module named 'app.models.orm.version_metadata'`

File: app/crud/__init__.py
Comment (solomon-negusse) [reply app/crud/__init__.py:19]: Good catch, it seems like it! Will leave it to another PR to test removing it.

File: app/crud/assets.py
Comment (solomon-negusse) [reply app/crud/assets.py:171]: Not sure I follow..are you referring to how we pass objects thru this json encoder in many places?

**Comment by solomon-negusse** [comment]
> I added a few small comments in here, @solomon-negusse. This PR isn't as big as it looks because of all the other merged commits from other efforts. 😅
> 
> I see that the tests are failing, still. Have you been doing something to get them to pass in your environment? The error is : `ModuleNotFoundError: No module named 'app.models.orm.version_metadata'`

@gtempus, just heads up I pushed a small fix to `metadata` branch if you have pulled it for testing. Thanks again for the review.

File: app/crud/metadata.py
Comment (jterry64) [inline app/crud/metadata.py:33]: I see this query a few times, should it just be a helper function?

File: app/models/pydantic/jobs.py
Comment (jterry64) [inline app/models/pydantic/jobs.py:133]: Is this supposed to be part of your PR? Seems unrelated

File: app/routes/analysis/analysis.py
Comment (jterry64) [inline app/routes/analysis/analysis.py:8]: Do we want to disable API keys in staging again, or is that on accident?

File: app/routes/assets/asset.py
Comment (jterry64) [inline app/routes/assets/asset.py:269]: I don't think it's super necessary, but it feels weird you can PATCH a specific field, but you can't do a GET on a specific field (only all fields). Should we include it for consistency?

File: app/routes/datasets/downloads.py
Comment (jterry64) [inline app/routes/datasets/downloads.py:8]: Same Q as above

File: app/routes/datasets/downloads.py
Comment (jterry64) [inline app/routes/datasets/downloads.py:41]: Same as above as well - is this supposed to be part of this PR, or is this merged from a different PR?

File: app/routes/datasets/versions.py
Comment (jterry64) [inline app/routes/datasets/versions.py:134]: So this just throws it away for legacy versions? Unclear what the purpose of doing this is

File: app/tasks/raster_tile_cache_assets/symbology.py
Comment (jterry64) [inline app/tasks/raster_tile_cache_assets/symbology.py:413]: This is part of this PR or from a PR from @dmannarino?

**Review by jterry64 — CHANGES_REQUESTED** [review]
Looks pretty good, just some minor comments! Mainly, it seems like there might be some other changes related to geostore/tile caches that accidentally got merged into here, just wanted to confirm (or maybe they're just already in staging but not in your metadata-final branch?)

File: app/routes/assets/asset.py
Comment (solomon-negusse) [reply app/routes/assets/asset.py:269]: Good catch, I'm actually leaning to modifying the PATCH to update any number of fields.

File: app/routes/datasets/versions.py
Comment (solomon-negusse) [reply app/routes/datasets/versions.py:134]: This is to avoid sending version metadata which is no longer valid for assets to `create_default_asset` function in the next step which will try to process any incoming `metadata` in `input_data`.

File: app/routes/analysis/analysis.py
Comment (solomon-negusse) [reply app/routes/analysis/analysis.py]: Yes, disabling them for now until Vizz devs are ready to test it.

File: app/models/pydantic/jobs.py
Comment (solomon-negusse) [reply app/models/pydantic/jobs.py]: As you figured in other comment there was a mismatch in the base of these two branches: `metadata-final` was based on `develop` and `metadata-endpoints-crud` was based on master which had a lot of hot fixes. I have fixed that now.

File: app/crud/metadata.py
Comment (solomon-negusse) [reply app/crud/metadata.py:33]: Hmm..it's one line without the `black` line wrapping:) but we should do that if these functions grow in size.

File: app/crud/datasets.py
Comment (solomon-negusse) [reply app/crud/datasets.py:27]: Gino syntax for relational query is less intuitive than in sqlalchemy:/ I have added a comment in the code to clarify it.

File: app/crud/assets.py
Comment (solomon-negusse) [reply app/crud/assets.py:231]: Ah that makes sense. That pattern is used the API as a whole. Can we work on that in a future refactor iteration given that the timeline for metadata is tight now?

Comment (gtempus) [reply app/crud/assets.py:231]: Totes! It was just an observation. We can consider this in later changes, @solomon-negusse.

File: app/crud/versions.py
Comment (dmannarino) [inline app/crud/versions.py:51]: This is fine, but would it be just as well (and slightly more succinct) to do
row.metadata = getattr(row, "metadata", {})
? Or could metadata be None anyway?

File: app/routes/assets/asset.py
Comment (dmannarino) [inline app/routes/assets/asset.py:338]: Would HTTP 409 be better here?

File: app/routes/datasets/versions.py
Comment (dmannarino) [inline app/routes/datasets/versions.py:396]: I think 409 would be better

File: app/crud/versions.py
Comment (solomon-negusse) [reply app/crud/versions.py]: Duh, updated.  The pydantic model not liking `None` so updated with `{}` default.

**Review by dmannarino — APPROVED** [review]

**Comment by dmannarino** [comment]
Looks good to me! However, would you mind waiting until after we merge dev -> master tomorrow morning with some unrelated changes?

**Comment by solomon-negusse** [comment]
> Looks good to me! However, would you mind waiting until after we merge dev -> master tomorrow morning with some unrelated changes?

Yeah, it can wait.

**Pull request closed without merging**
